src/metabase/api/api_key.clj

```diff
@@ -29,28 +30,83 @@
   (api/check-superuser)
   (api/checkp (not (t2/exists? :model/ApiKey :name name))
     "name" "An API key with this name already exists.")
-  (let [api-key (key-with-unique-prefix)
-        email   (format "api-key-user-%s@api-key.invalid" (u/slugify name))]
+  (let [unhashed-key (key-with-unique-prefix)
+        email        (format "api-key-user-%s@api-key.invalid" (u/slugify name))]
     (t2/with-transaction [_conn]
       (let [user (user/insert-new-user! {:email      email
                                          :first_name name
                                          :type       :api-key})]
         (user/set-permissions-groups! user [(perms-group/all-users) group_id])
-        (-> (t2/insert-returning-instance! :model/ApiKey
-                                           {:user_id      (u/the-id user)
-                                            :name         name
-                                            :unhashed_key api-key
-                                            :created_by   api/*current-user-id*})
-            (select-keys [:created_at :updated_at :id])
-            (assoc :name name
-                   :group_id group_id
-                   :unmasked_key api-key
-                   :masked_key (api-key/mask api-key)))))))
+        (let [api-key (-> (t2/insert-returning-instance! :model/ApiKey
+                                                         {:user_id      (u/the-id user)
+                                                          :name         name
+                                                          :unhashed_key unhashed-key
+                                                          :created_by   api/*current-user-id*})
+                          (t2/hydrate :group_name))]
+          (events/publish-event! :event/api-key-create
+                                 {:object  api-key
+                                  :user-id api/*current-user-id*})
+          (-> api-key
+              (select-keys [:created_at :updated_at :id :group_name :name])
+              (assoc :unmasked_key unhashed-key
+                     :masked_key (api-key/mask unhashed-key))))))))
 
 (api/defendpoint GET "/count"
   "Get the count of API keys in the DB"
   [:as _body]
   (api/check-superuser)
   (t2/count :model/ApiKey))
 
+(api/defendpoint PUT "/:id"
+  "Update an API key by changing either its group or its name"
+  [id :as {{:keys [group_id name] :as _body} :body}]
+  {id       ms/PositiveInt
+   group_id [:maybe ms/PositiveInt]
+   name     [:maybe ms/NonBlankString]}
+  (api/check-superuser)
+  (let [api-key-before (-> (t2/select-one :model/ApiKey :id id)
+                           ;; hydrate the group_name for audit logging
+                           (t2/hydrate :group_name))]
+    (when name
+      (t2/with-transaction [_conn]
```

Comment:
If `name` and `group_id` are changed in the same call, then maybe the same transaction should be used for both updates